### PR TITLE
Animation

### PR DESCRIPTION
**Pull request by NOOBDY**

### CODE REVIEW

File: src/Core/Context.cpp [hunk 1]
```diff
@@ -102,7 +102,13 @@ void Context::Update() {
     Util::Input::Update();
     SDL_GL_SwapWindow(m_Window);
     glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+    if constexpr (FPS_CAP != 0) {
+        constexpr double frameTime = 1000 / static_cast<double>(FPS_CAP);
+        SDL_Delay(static_cast<Uint32>(frameTime - Util::Time::GetDeltaTime()));
+    }
 }
```
Comment (Error0229) [inline src/Core/Context.cpp:110]: The `Util::Time::GetDeltaTime()` is getting the delta time between "last frame start time" and "current frame start time", Which is not the delta time between "current frame start" and the moment "this function got called". Which cause it always Delay for wrong period.

File: include/Util/Animation.hpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    enum class State {
+        PLAY,  /**< Animation is playing. */
+        PAUSE, /**< Animation is paused. */
+    };
+
+    /**
+     * @brief Constructor for Animation class.
+     * @param paths Vector of file paths for the frames.
+     * @param bool Whether the animation should play right away.
```
Comment (Error0229) [inline include/Util/Animation.hpp:30]: `@param play`

**Review by Error0229 — COMMENTED** [review]
+ Some debug message may cleared.
+ Animation state's getter aren't complete.
+ However most of given API works fine.

File: src/Util/Animation.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+void Animation::Update() {
+    if (m_State == State::PAUSE) {
+        LOG_TRACE("[ANI] is pause");
+        m_LastFrameStartTime = Util::Time::GetElapsedTimeMs();
+        return;
+    }
+
+    if (!m_Looping && m_HasEnded) {
+        LOG_TRACE("[ANI] not loop and is ended");
+        m_LastFrameStartTime = Util::Time::GetElapsedTimeMs();
```
[the remark this answers is not in the document]
Comment (NOOBDY) [reply src/Util/Animation.cpp:36]: it gets the job done, also I can't come up with a better name

File: src/Util/Animation.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            if (delta > m_Interval + m_Cooldown) {
+                addFrameCount = (delta - m_Cooldown) / m_Interval;
+
+                m_LastFrameStartTime += addFrameCount * m_Interval + m_Cooldown;
+            }
+        } else if (delta > m_Interval) {
+            addFrameCount = delta / m_Interval;
+
+            m_LastFrameStartTime += addFrameCount * m_Interval;
+        }
```
[the remark this answers is not in the document]
Comment (NOOBDY) [reply src/Util/Animation.cpp]: i feel like the implementation here is fucked

File: include/Util/Time.hpp [hunk 5]
```diff
@@ -17,6 +17,8 @@ namespace Util {
  */
 class Time {
 public:
+    static void StartCounter() { s_Start = SDL_GetPerformanceCounter(); }
```
Comment (Dragonfly911117) [inline include/Util/Time.hpp:20]: Need docs for this method. Also, what the purpose of this method? `s_Start` already inited with `SDL_GetPerformanceCounter()`.

File: include/Util/Time.hpp [hunk 6]
```diff
@@ -27,6 +29,8 @@ class Time {
      */
     static double GetDeltaTime() { return s_DeltaTime; }
 
+    static unsigned long GetElapsedTimeMs();
```
Comment (Dragonfly911117) [inline include/Util/Time.hpp:32]: Need docs for this method.

Comment (Dragonfly911117) [reply include/Util/Time.hpp:32]:
```suggestion
    /**
     * @brief Get the elapsed time from the start of the program in
     * milliseconds.
     *
     * @note To create a timer, one may call this function to record the time,
     * and then call it again to obtain the time difference.
     *
     * @return The elapsed time from the start of the program in milliseconds.
     */
    static unsigned long GetElapsedTimeMs();```

File: src/Core/Context.cpp [hunk 7]
```diff
@@ -102,7 +102,13 @@ void Context::Update() {
     Util::Input::Update();
     SDL_GL_SwapWindow(m_Window);
     glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+    if (FPS_CAP != 0) {
+        constexpr double frameTime = 1000 / static_cast<double>(FPS_CAP);
+        SDL_Delay(static_cast<Uint32>(frameTime - Util::Time::GetDeltaTime()));
+    }
```
Comment (Dragonfly911117) [inline src/Core/Context.cpp:109]: Having `FPS_CAP` as `0` triggers "Constexpr variable 'frameTime' must be initialized by a constant expression division by zero". error. 
```suggestion
constexpr double frameTime =
        FPS_CAP != 0 ? 1000 / static_cast<double>(FPS_CAP) : 0;
    SDL_Delay(static_cast<Uint32>(frameTime - Util::Time::GetDeltaTime()));
```

**Review by Dragonfly911117 — CHANGES_REQUESTED** [review]
Seems good to me except some minor changes needed.

**Comment by Dragonfly911117** [comment]
Cool GitHub

**Review by Dragonfly911117 — APPROVED** [review]

**Pull request merged**